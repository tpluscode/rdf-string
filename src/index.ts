export { sparql } from './lib/sparql'
export { turtle } from './lib/turtle'
export { nquads } from './lib/nquads'
export { ntriples } from './lib/ntriples'
export type { SparqlTemplateResult, SparqlValue } from './lib/sparql'
export type { TurtleTemplateResult, TurtleValue } from './lib/turtle'
export type { NQuadsTemplateResult, NQuadsValue } from './lib/nquads'
export type { NTriplesTemplateResult, NTriplesValue } from './lib/ntriples'
