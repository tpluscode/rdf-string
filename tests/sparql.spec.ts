import { sparql } from '../src/index'
import { namedNode } from '@rdfjs/data-model'
import { prefixes } from '@zazuko/rdf-vocabularies'
import namespace from '@rdfjs/namespace'

const schema = namespace(prefixes.schema)
const foaf = namespace(prefixes.foaf)

describe('sparql', () => {
  it('serializes named node', () => {
    // given
    const type = namedNode('http://example.com/type')
    const expected = 'SELECT * WHERE { ?person a <http://example.com/type> }'

    // when
    const query = sparql`SELECT * WHERE { ?person a ${type} }`

    // then
    expect(query.toString()).toMatchQuery(expected)
  })

  it('serializes string literal', () => {
    // given
    const name = 'John Doe'
    const expected = 'SELECT * WHERE { ?person <http://schema.org/name> "John Doe" }'

    // when
    const query = sparql`SELECT * WHERE { ?person <http://schema.org/name> ${name} }`

    // then
    expect(query).toMatchQuery(expected)
  })

  it('extracts known prefixes', () => {
    // given
    const expected = `PREFIX schema: <http://schema.org/>
    SELECT * WHERE { ?person schema:name "Tomasz" }`

    // when
    const query = sparql`SELECT * WHERE { ?person ${schema.name} "Tomasz" }`

    // then
    expect(query).toMatchQuery(expected)
  })

  it('ignores null', () => {
    // given
    const expected = 'SELECT * WHERE { ?person <http://schema.org/name> "Tomasz" }'

    // when
    const query = sparql`SELECT * WHERE { ?person <http://schema.org/name> "Tomasz" ${null} }`

    // then
    expect(query).toMatchQuery(expected)
  })

  it('ignores undefined', () => {
    // given
    const expected = 'SELECT * WHERE { ?person <http://schema.org/name> "Tomasz" }'

    // when
    const query = sparql`SELECT * WHERE { ?person <http://schema.org/name> "Tomasz" ${undefined} }`

    // then
    expect(query).toMatchQuery(expected)
  })

  it('merges nested templates, hoisting prefixes', () => {
    // given
    const expected = `
    PREFIX schema: <http://schema.org/>
    PREFIX foaf: <http://xmlns.com/foaf/0.1/>
    
    SELECT * WHERE {
     ?person schema:name "Tomasz" .
     ?person foaf:knows ?friend .
    }`

    // when
    const namePattern = sparql`?person ${schema.name} "Tomasz" .`
    const knowsPattern = sparql`?person ${foaf.knows} ?friend .`
    const query = sparql`SELECT * WHERE { ${namePattern} ${knowsPattern} }`

    // then
    expect(query).toMatchQuery(expected)
  })
})
