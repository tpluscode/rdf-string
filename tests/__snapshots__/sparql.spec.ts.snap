// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sparql interpolating named node escapes slashes and hash in prefixed names 1`] = `"SELECT * WHERE { ex::foo\\\\/bar\\\\#baz <eats> ex::foo\\\\/bar\\\\#baz }"`;

exports[`sparql interpolating named node respects base in sub-templates 1`] = `
"BASE <http://example.org/>

SELECT * WHERE { <dog> <eats> <dog> }"
`;

exports[`sparql interpolating named node writes URIs relative to base 1`] = `
"BASE <http://example.org/>

PREFIX : <http://example.org/vocab#> 
SELECT * WHERE { <dog> :eats <dog> }"
`;
